beep/features/knee_identification.py

```diff
@@ -0,0 +1,105 @@
+    
+    Inputs
+    seq_num (int): sequence number of battery cell, unique identifier
+    batch_df (DataFrame): df with index as seq_num, columns 'fractional_metric' and 'equivalent_full_cycles'
+    
+    Outputs
+    knee_df: one-row dataframe containing values that describe the knee behavior
+         a0: y-value of initial capacity point
+         af: y-value of final capacity point
+         xt: x-value of transition between the two fitted linear lines
+         yt: y-value of transition between the two fitted linear lines
+         xf: set to 1 to find slope over normalized (0-1) axes
+   xf_scale: x-value of final capacity value, used to rescale to full scale
+ slope_diff: difference in the value of slope between the first and second fitted lines
+ angle_diff: difference in value of angle (radians) between the first and second fitted lines
+      error: squared error during the fitting routine. Values less than 1e-3 are typically acceptable
+    """
+    degradation_path_seq_num = batch_df.loc[seq_num]
+    degradation_path_seq_num = degradation_path_seq_num.loc[degradation_path_seq_num['fractional_metric'] > threshold]
+    
+    a0 = degradation_path_seq_num['fractional_metric'].iloc[0]
+    af = np.min(degradation_path_seq_num['fractional_metric'])
+    xf = 1
+    xf_scale = np.max(degradation_path_seq_num['equivalent_full_cycles'])
+    bacon_watts_param_bounds = ((degradation_path_seq_num['fractional_metric'].iloc[0],degradation_path_seq_num['fractional_metric'].iloc[0]), # a0
+                                (np.min(degradation_path_seq_num['fractional_metric']),np.min(degradation_path_seq_num['fractional_metric'])), # af
+                                (0,1), # xt
+                                (af,1), # yt
+                                (1,1)) # xf
+    opt_result = differential_evolution(get_error_bacon_watts_heaviside,
+                                                 bacon_watts_param_bounds,
+                                                 args=(degradation_path_seq_num['fractional_metric'],degradation_path_seq_num['equivalent_full_cycles']),
+                                                 strategy='best1bin', maxiter=10000,
+                                                 popsize=500, tol=0.1e-8, mutation=0.5, recombination=0.7, seed=1,
+                                                 callback=None, disp=None, polish=True, init='latinhypercube', atol=1e-7, updating='deferred', workers=-1, constraints=())
+    xt = opt_result.x[2]
+    yt = opt_result.x[3]
+
+    m1 = (yt-a0)/(xt)
+    m2 = (af-yt)/(xf-xt)
+    slope_diff = m1-m2
+    angle_diff = np.arctan(m1) - np.arctan(m2)
+    
+    error = opt_result.fun
+    
+    knee_df = pd.DataFrame([[a0, af, xt, yt, xf, xf_scale, slope_diff, angle_diff, error]],
+                           columns=['a0', 'af', 'xt', 'yt', 'xf', 'xf_scale', 'slope_diff', 'angle_diff', 'error'])
+    return knee_df
```

Comment:
Some minor formatting errors are cropping up during linting:
5       E127 continuation line over-indented for visual indent
20      E231 missing whitespace after ','
5       E261 at least two spaces before inline comment
3       E302 expected 2 blank lines, found 1
6       E501 line too long (155 > 120 characters)
1       W292 no newline at end of file